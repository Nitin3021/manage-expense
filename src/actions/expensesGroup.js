import moment from 'moment'
import database from '../firebase/firebase'

export const addExpenseGroup = (expensesGroup) => ({
    type: 'ADD_EXPENSES_GROUP',
    expensesGroup
})

export const startAddExpenseGroup = (expensesGroupData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            description = ''
        } = expensesGroupData
        const expensesGroup = { 
            description,
            createdAt: moment().valueOf()
         }

        return database.ref(`users/${uid}/expensesGroup`).push(expensesGroup).then((ref) => {
            dispatch(addExpenseGroup({
                id: ref.key,
                ...expensesGroup
            }))
        })
    }
}

export const setExpensesGroup = (expensesGroup) => ({
    type: 'SET_EXPENSES_GROUP',
    expensesGroup
})

export const startSetExpensesGroup = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return database.ref(`users/${uid}/expensesGroup`).once('value').then((snapshot) => {
            const expensesGroup = []

            snapshot.forEach((childSnapshot) => {
                expensesGroup.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })

            dispatch(setExpensesGroup(expensesGroup))
        })
    }
}