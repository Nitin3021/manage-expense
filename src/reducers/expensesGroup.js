const expensesGroupDefaultState = []

export default (state = expensesGroupDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSES_GROUP':
            return [
                ...state,
                action.expensesGroup
            ]
        case 'SET_EXPENSES_GROUP':
            return action.expensesGroup
        default:
            return state
    }
};