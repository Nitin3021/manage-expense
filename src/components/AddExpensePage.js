import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import { toastIt } from '../actions/toasts'

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        const groupId = (window.location.href).split('/create/')[1];
        expense["groupId"] = groupId;

        this.props.startAddExpense(expense);
        this.props.toastIt(
            'add', 
            `${expense.description} added!`
        )
        this.props.history.push(`/dashboard/${groupId}`);
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
    toastIt: (actionType, text) => dispatch(toastIt(actionType, text))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);