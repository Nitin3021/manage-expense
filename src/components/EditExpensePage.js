import React from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import { toastIt } from '../actions/toasts';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.toastIt(
            'edit',
            `${this.props.expense.description} modified!`
        )
        this.props.history.push(`/dashboard/${this.props.groupId}`);
    };

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.toastIt(
            'remove',
            `${this.props.expense.description} removed!`
        )
        this.props.history.push(`/dashboard/${this.props.groupId}`);
    };

    onButtonClick = () => {
        this.props.history.push(`/dashboard/${this.props.groupId}`);
    }

    checkLink = () => {
        if (this.props.expensesGroupMatch === undefined) {
            return <Redirect to='/InvalidId' />
        }
    }

    render() {
        return (
            <div className="box-layout_expensepage">
                {this.checkLink()}
                <div>
                    <div className="page-header">
                        <div className="content-container">
                            <div className="box_layout_expensepageheader">
                                <h1 className="page-header__title">Edit Expense</h1>
                                <button className="button button--expensepage" onClick={this.onButtonClick}>Back</button>
                            </div>
                        </div>
                    </div>
                    <div className="content-container">
                        <ExpenseForm
                            expense={this.props.expense}
                            onSubmit={this.onSubmit}
                        />
                        <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
                    </div>
                </div>
            </div>
        );
    };
};

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
    groupId: state.toasts.groupId,
    expensesGroupMatch: state.expensesGroup.find((expenseGroup) => expenseGroup.id === state.toasts.groupId)
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (groupId, id, expense) => dispatch(startEditExpense(groupId, id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
    toastIt: (actionType, text) => dispatch(toastIt(actionType, text))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);