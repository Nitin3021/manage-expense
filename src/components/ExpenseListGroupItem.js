import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { connect } from 'react-redux';
import { startRemoveExpense } from '../actions/expenses';
import { startRemoveExpensesGroup } from '../actions/expensesGroup';

const ExpenseListGroupItem = (props) => {
    const onRemoveGroup = (e) => {
        e.preventDefault()

        const expensesFiltered = props.expenses.filter((expense) => expense.groupId === props.id)

        expensesFiltered.forEach(expense => {
            props.startRemoveExpense({ id: expense.id })
        });

        props.startRemoveExpensesGroup({ id: props.id })
        window.location.reload()
    }

    return (
        <Link className="list-item" to={`/dashboard/${props.id}`}>
            <div>
                <h3 className="list-item__title">{props.description}</h3>
                <span className="list-item__sub-title">{moment(props.createdAt).format('Do MMMM YYYY')}</span>
                <button onClick={onRemoveGroup}>Delete</button>
            </div>
        </Link>
    )
}

const mapStateToProps = (state) => ({
    expenses: state.expenses
})

const mapDispatchToProps = (dispatch) => ({
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data)),
    startRemoveExpensesGroup: (data) => dispatch(startRemoveExpensesGroup(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListGroupItem);