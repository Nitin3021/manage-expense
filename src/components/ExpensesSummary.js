import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import { toastGroup } from '../actions/toasts';

export const ExpensesSummary = (props) => {
    const expenseWord = props.expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(props.expensesTotal / 100).format('$0,0.00');

    useEffect(() => {
        props.toastGroup(props.groupId)
    })

    const checkLink = () => {
        if (props.expensesGroupMatch === undefined) {
            return <Redirect to='/InvalidId' />
        }
    }

    return (
        <div className="page-header">
            {checkLink()}
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{props.expenseCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span> </h1>
                <div className="page-header__actions">
                    <Link className="button" to={`/create/${props.groupId}`}>Add Expense</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    const groupId = (window.location.href).split('/dashboard/')[1];
    const visibleExpenses = selectExpenses(groupId, state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses),
        expensesGroupMatch: state.expensesGroup.find((expenseGroup) => expenseGroup.id === groupId),
        groupId
    };
};

const mapDispatchToProps = (dispatch) => ({
    toastGroup: (groupId) => dispatch(toastGroup(groupId))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpensesSummary);