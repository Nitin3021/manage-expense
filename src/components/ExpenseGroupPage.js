import React from "react"
import { connect } from 'react-redux'
import ExpenseListGroupItem from './ExpenseListGroupItem'
import { startAddExpenseGroup } from '../actions/expensesGroup'

export class ExpenseGroupPage extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            description: props.expensesGroup.length !== 0 ? props.expensesGroup.description : ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }

    onSubmit = (e) => {
        e.preventDefault()
        this.props.startAddExpenseGroup({ description: this.state.description })
    }

    render() {
        return (
            <div className="content-container">
                <form className="form" onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Expense Group Name"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />

                    <button>Add Expense Group</button>
                </form>

                {
                    this.props.expensesGroup.length === 0 ? (
                        <div className="list-item list-item--message">
                            <span>No Expenses!</span>
                        </div>
                    ) : (
                        this.props.expensesGroup.map((expenseGroup) => {
                            return <ExpenseListGroupItem key={expenseGroup.id} {...expenseGroup} />
                        })
                    )
                }
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    expensesGroup: state.expensesGroup
})

const mapDispatchToProps = (dispatch) => ({
    startAddExpenseGroup: (expensesGroupData) => dispatch(startAddExpenseGroup(expensesGroupData))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseGroupPage)