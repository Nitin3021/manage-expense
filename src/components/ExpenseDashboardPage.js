import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
import ToasterMsg from './ToasterMsg';

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList />
        <ToasterMsg />
    </div>
);

export default ExpenseDashboardPage;