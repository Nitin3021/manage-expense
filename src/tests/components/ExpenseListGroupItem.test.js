import React from 'react';
import { shallow } from 'enzyme';
import expensesGroup from '../fixtures/expensesGroup';
import { ExpenseListGroupItem } from '../../components/ExpenseListGroupItem';

test('Should render ExpenseListGroupItem correctly', () => {
    const wrapper = shallow(<ExpenseListGroupItem {...expensesGroup[0]} />);
    expect(wrapper).toMatchSnapshot();
})