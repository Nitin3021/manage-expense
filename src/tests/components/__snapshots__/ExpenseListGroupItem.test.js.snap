// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseListGroupItem correctly 1`] = `
<div>
  <Link
    className="list-item"
    to="/dashboard/1"
  >
    <div>
      <h3
        className="list-item__title"
      >
        Home Expense
      </h3>
      <span
        className="list-item__sub-title"
      >
        1st January 1970
      </span>
    </div>
    <img
      className="list-item__deleteButton"
      onClick={[Function]}
      src="/images/delete.png"
    />
  </Link>
  <ModalDeletion
    handleClearSelectedOption={[Function]}
    onRemoveGroup={[Function]}
  />
</div>
`;
